curp/src/client.rs

```diff
@@ -221,6 +218,51 @@ pub enum ReadState {
     CommitIndex(LogIndex),
 }
 
+/// Unpack `tonic::Status`
+enum UnpackStatus {
+    /// CurpServer is shutting down
+    ShuttingDown,
+    /// Indicate that the client sent a wait synced request to a non-leader
+    Redirect(Option<ServerId>, u64),
+    /// Transport error type
+    Transport,
+    /// Encode decode error type
+    EncodeDecode,
+    /// Storage error type
+    Storage,
+    /// IO error type
+    IO,
+    /// Internal error type
+    Internal,
+}
+
+/// unpack `tonic::Status` and convert it to `UnpackStatus`
+fn unpack_status(status: &tonic::Status) -> UnpackStatus {
+    let meta = status.metadata();
+    if let Some(label) = meta.get("error-label") {
+        match label.to_str().unwrap_or_else(|err| {
+            unreachable!("error-label should be always able to convert to str: {err:?}")
+        }) {
+            "shutting-down" => UnpackStatus::ShuttingDown,
+            "transport" => UnpackStatus::Transport,
+            "redirect" => {
+                let (leader_id, term) = serde_json::from_slice(status.details()).unwrap_or_else(|err| unreachable!(" deserialize (leader_id, term) from status' detail should always success: {err:?}"));
+                UnpackStatus::Redirect(leader_id, term)
+            }
+            "encode-decode" => UnpackStatus::EncodeDecode,
+            "internal" => UnpackStatus::Internal,
+            "storage" => UnpackStatus::Storage,
+            "io" => UnpackStatus::IO,
+            unsupported_label => {
+                unreachable!("unsupported status label {unsupported_label}")
+            }
+        }
+    } else {
+        // This transport error comes from `tonic` framework
+        UnpackStatus::Transport
+    }
+}
```

Comment:
Shoude we use `From` trait here?

Reply:
The From trait will consume the `Status` value and the `Status` value is still needed in some other methods, like `slow_round`